text/0000-module-unification.md

```diff
@@ -0,0 +1,829 @@
+possible that we may allow overrides / aliases in the future.
+
+Modules that appear alongside `main.js` in `src` are also considered `main`
+modules for their respective `type`. For instance, `src/router.js` is registered
+with a `name` of `main` and a `type` of `router`.
+
+### Module Collections
+
+Top-level namespaces within `src` serve to group modules into
+type-aware "collections".
+
+The following rules apply to module collections and types:
+
+1. Each collection can contain one or more types. The types allowed
+   in a particular collection MUST be explicitly declared.
+2. Each type MAY exist in any number of collections.
+3. Each type MUST have only one "definitive collection", which is the
+   collection the resolver will use for resolutions if a module can't be found
+   in the local (i.e. originating) collection.
+4. Each collection MAY have a single "default type". If a module does not
+   indicate its type through its file name, then its default export should
+   align with the default type for its collection.
+5. Each collection can allow "private collections" to be defined at a namespace.
+   Private collections are localized additions to a top-level collection,
+   available only from the namespace at which they're defined.
+6. Top-level collections may be grouped for organization purposes. No
+   resolvable modules must be placed in a group directory.
+7. A collection can appear only once in a project (i.e. it can not be
+   contained in multiple group directories, or in a group as well as at the
+   top-level).
+
+The following collections and allowed types (rules 1 & 2) are proposed:
+
+* `components` - COMPONENT, HELPER, template
+* `initializers` - INITIALIZER
+* `instance-initializers` - INSTANCE-INITIALIZER
+* `models` - MODEL, ADAPTER, SERIALIZER
+* `partials` - PARTIAL
+* `routes` - ROUTE, CONTROLLER, template
+* `services` - SERVICE
+* `transforms` - TRANSFORM
+* `utils` - UTIL
+
+> Note: ALL CAPS indicates which collections are definitive (rule 3) for a type.
+
+The following default types are proposed for collections (rule 4):
+
+* `components` - component
+* `initializer` - initializer
+* `instance-initializers` - instance-initializer
+* `models` - model
+* `partials` - partial
+* `routes` - route
+* `services` - service
+* `transforms` - transform
+* `utils` - util
+
+The following private collections are allowed within collections (rule 5):
+
+* `components` - utils
+* `models` - utils
+* `initializers` - utils
+* `instance-initializers` - utils
+* `routes` - components, utils
+* `services` - utils
+* `transforms` - utils
+
+The following groups are proposed for collections (rule 6):
+
+* `data` - models, transforms
+* `init` - initializers, instance-initializers
+* `ui` - components, partials, routes
+
+The collection and type system is designed to be extensible, so that addons can
+contribute their own collections and types. The `data` collection and its
```

Comment:
Yes, assuming `fragment` was not already reserved by something else, addons will be able to add types to existing collections.